lib/model/autocomplete.dart

```diff
@@ -461,3 +514,78 @@ class UserMentionAutocompleteResult extends MentionAutocompleteResult {
 // TODO(#233): // class UserGroupMentionAutocompleteResult extends MentionAutocompleteResult {
 
 // TODO(#234): // class WildcardMentionAutocompleteResult extends MentionAutocompleteResult {
+
+class TopicAutocompleteView extends AutocompleteView<TopicAutocompleteQuery, TopicAutocompleteResult, String> {
+  TopicAutocompleteView._({required super.store, required this.streamId});
+
+  factory TopicAutocompleteView.init({required PerAccountStore store, required int streamId}) {
+    final view = TopicAutocompleteView._(store: store, streamId: streamId);
+    store.autocompleteViewManager.registerTopicAutocomplete(view);
+    view._fetch();
+    return view;
+  }
+
+  final int streamId;
+  Iterable<String> _topics = [];
+  bool _isFetching = false;
+
+
+  /// Fetches topics of the current stream narrow, expected to fetch
+  /// only once per lifecycle.
+  ///
+  /// Starts fetching once the stream narrow is active, then when results
+  /// are fetched it restarts search to refresh UI showing the newly
+  /// fetched topics.
+  Future<void> _fetch() async {
+    if (_isFetching) return;
+    _isFetching = true;
+    final result = await getStreamTopics(store.connection, streamId: streamId);
+    _topics = result.topics.map((e) => e.name);
+    _isFetching = false;
+    if (_query != null) _startSearch(_query!);
+  }
+
+  @override
+  Iterable<String> getSortedItemsToTest(TopicAutocompleteQuery query) => _topics;
+
+  @override
+  TopicAutocompleteResult? testItem(TopicAutocompleteQuery query, String item) {
+    if (query.testTopic(item)) {
+      return TopicAutocompleteResult(topic: item);
+    }
+    return null;
+  }
+
+  @override
+  void dispose() {
+    store.autocompleteViewManager.unregisterTopicAutocomplete(this);
+    super.dispose();
+  }
+}
+
+class TopicAutocompleteQuery extends AutocompleteQuery {
+  TopicAutocompleteQuery(super.raw);
+
+  bool testTopic(String topic) => topic.isNotEmpty
+    && topic != raw
+    && topic.toLowerCase().contains(raw.toLowerCase());
```

Comment:
Yes, this is a followup to [#627 (comment)](https://github.com/zulip/zulip-flutter/pull/627#discussion_r1697748918), unfortunately I could not load the comment when I worked on this revision due to the large number of comments which caused page to timeout without loading all comments..

And yes, I believe this copies what [zulip-mobile does](https://github.com/zulip/zulip-mobile/blob/main/src/autocomplete/TopicAutocomplete.js#L59)  and it also matches what [web does](https://github.com/zulip/zulip/blob/main/web/src/composebox_typeahead.ts#L1216)

Reply:
Great, thanks for those references.

It looks like this does match what zulip-mobile does.

Let's drop the `isNotEmpty` condition — it doesn't make sense to me why that should be there. Looking in the Git history (`git log -L 59,61:src/autocomplete/TopicAutocomplete.js`), it was added in https://github.com/zulip/zulip-mobile/pull/1478 with no explanation. I suspect the point there was actually to check `x` wasn't `null` or `undefined`, rather than that it wasn't the empty string; back in 2017 that codebase didn't really have type-checking, so a null or undefined value could easily slip in where it didn't belong. (Even then, the right fix would have been to stop the unwanted value from entering in the first place, rather than paper over it like that commit did.)

Let's also drop the `topic != raw` condition. It doesn't seem to match the behavior of web.

Looking at web's code, it also has some reordering among the matches. Let's leave that for a future, post-launch issue, though. … OK, I've filed #881 for that — so we can leave here a comment like `// TODO(#881): sort by match relevance`.